README.md
<!-- badges: start -->
[![R-CMD-check](https://github.com/campbio/scruff/workflows/R-CMD-check/badge.svg)](https://github.com/campbio/scruff/actions)
<!-- badges: end -->

# scruff: Single Cell RNA-Seq UMI Filtering Facilitator

[**scruff**](https://doi.org/10.1186/s12859-019-2797-2) (**S**ingle **C**ell **R**NA-Seq **U**MI **F**iltering **F**acilitator) is a package for processing single cell RNA-seq (scRNA-seq) FASTQ reads generated by CEL-Seq and CEL-Seq2 protocols. It demultiplexes scRNA-seq FASTQ files, aligns reads to reference genome using [Rsubread](https://doi.org/10.1093/nar/gkz114), and generates UMI filtered count matrix.

## Workflow
<p align="center"><img src="https://github.com/campbio/scruff/raw/master/data-raw/figure/20190312_scruff_workflow.png" height="900"></p>

To install the latest stable release of `scruff` from [Bioconductor](http://bioconductor.org/packages/scruff/):
```
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("scruff")
```

To install the development version of `scruff` from GitHub using `devtools`:
```
library(devtools)
install_github("campbio/scruff")
```

## Package User Guide

An introduction to `scruff` package is available [here](http://bioconductor.org/packages/release/bioc/vignettes/scruff/inst/doc/scruff.html).

## Speed

<p align="center"><img src="https://github.com/campbio/scruff/raw/master/data-raw/figure/20190312_scruff_figure_5.png" height="400"></p>

Run time benchmarks for selected scRNA-seq preprocessing packages. FASTQ files from the example dataset [(Van den Brink et al. 2017)](https://www.nature.com/articles/nmeth.4437) were subsampled to have a total read number of 0.1, 0.5, 1.0, 5.0, and 10.0 million. Each of the subsampled datasets was processed by [celseq2](https://github.com/yanailab/celseq2), [scPipe](https://doi.org/10.1371/journal.pcbi.1006361), and **scruff**. All 3 packages were parallelized with 16 cores. The job was run on a cluster node with 2 eight-core 2.6 GHz Intel Xeon E5-2670 CPUs and 256 GB memory. For the subsampled dataset with 10 million reads, scruff took 34.9 minutes to finish while celseq2 and scPipe took 143.2 and 69.0 minutes.

## Example QC plots

The following selected QC plots are generated using data from [(Van den Brink et al. 2017)](https://www.nature.com/articles/nmeth.4437) and **scruff** package. Metrics and QC plots reported by **scruff** includes total number of reads, number of reads mapped to reference genome, number of reads mapped to genes, fraction of mapped reads to total reads, fraction of reads mapped to genes to reads mapped to genome, fraction of reads mapped to genes to total number of reads, total number of transcripts, number of mitochondrial transcripts, fraction of mitochondrial transcripts, number of transcribed genes, fraction of protein coding genes, fraction of protein coding transcripts, median and average number of reads per corrected and uncorrected UMI counts, and the number of detected genes divided by total number of reads sequenced per million.

![scruff total_reads](https://github.com/campbio/scruff/raw/master/data-raw/figure/20180907_vdb_newplots_ercc_edit_Page_01.png)

![scruff aligned_reads_fraction](https://github.com/campbio/scruff/raw/master/data-raw/figure/20180907_vdb_newplots_ercc_edit_Page_04.png)

These are boxplots with overlaid points. Each point represents a well (unique cell barcode) and is colored by the number of cells sorted in the well by FACS. Each boxplot denotes a different experiment (i.e. plate). Sample *mouse c library 2* have similar number of total reads but much less fraction of aligned reads compared to other experiments, showing its poor quality.

## Read alignment visualization colored by UMI

![read_alignment_UMI](https://github.com/campbio/scruff/raw/master/data-raw/figure/20190124_scruff_figure_3.png)

**scruff** package provides functions to visualize gene isoforms and UMI tagged read alignments at specified genomic coordinates. 125 reads were mapped to the gene Fos in cell 30 of mouse b library 1 from the example CEL-Seq dataset [(Van den Brink et al. 2017)](https://www.nature.com/articles/nmeth.4437). Upper panel shows the visualization of read alignments. Reads are represented by arrows and are colored by their UMIs. The direction of the arrow represents the mapping strand of the read. Lower panel shows the visualization of gene isoforms. Gene isoforms are labeled by their transcript names. Grey rectangles represent exons.

## QC plots for 10X Genomics data

![read_alignment_UMI](https://github.com/campbio/scruff/raw/master/data-raw/figure/20181221_scruff_figure_4.png)

**scruff** provides function to visualize read alignment quality of BAM files processed by Cell Ranger. BAM files for 6 PBMC datasets ([3K](https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/pbmc3k), [6K](https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/pbmc6k), [4K](https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/pbmc4k), [8K](https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/pbmc8k), [1K](https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_1k_v3), [10K](https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_10k_v3)) were downloaded from 10X Genomics website and processed by **scruff** to obtain (a) the number of reads aligned to reference genome, (b) the number of reads mapped to genes, and (c) the fraction of reads mapped to genes out of total number of aligned reads.

## How to cite [**scruff**](https://doi.org/10.1186/s12859-019-2797-2)
- Wang Z, Hu J, Johnson WE, Campbell JD. scruff: an R/Bioconductor package for preprocessing single-cell RNA-sequencing data. BMC bioinformatics. 2019;20(1):222.


NEWS.md
# scruff 1.19.1
* Refactored parallel prcessing using functions in parallelly and parallel

# scruff 1.10.1

* Fixed bugs related to using new readGFF function
* Added a `NEWS.md` file to track changes to the package.


R/gview.R
#' Visualize gene isoforms
#'
#' Visualize reference genome. Rectangles represent exons. Arrow represents
#'  orientation of transcripts.
#'
#' @param gtfFile A genome annotation file in GTF format.
#' @param chr Chromosome name. Integer or "X", "Y", "MT".
#' @param start Genomic coordinate of the start position.
#' @param end Genomic coordinate of the end position. If \code{NULL}, then
#' the maximum length of the chromosome will be used. Default \code{NULL}.
#' @param rect_width Exon widths. Default 0.3.
#' @param line_width Line weight. Default 0.5.
#' @param arrow_segments The number of segments lines be divided to. The
#'  greater the number, more arrows there are. Default 10.
#' @param arrow_width The angle of the arrow head in degrees (smaller numbers
#'  produce narrower, pointier arrows). Essentially describes the width of the
#'  arrow head. Passed to the angle parameter of arrow function. Default 30.
#' @param arrow_length The length of the arrow head. Passed to the length
#'  argument of arrow function. Default 0.08.
#' @param arrow_type One of "open" or "closed" indicating whether the arrow
#'  head should be a closed triangle. Passed to the type argument of arrow
#'  function. Default "open".
#' @param text_size Size of text. Passed to the size argument of the geom_text
#'  function. Default 4.
#' @return A ggplot object of genomic view
#' @examples
#' gtf <- system.file("extdata", "GRCm38_MT.gtf", package = "scruff")
#' g <- gview(gtf, chr = "MT")
#' g
#' @import rtracklayer
#' @export
gview <- function(gtfFile,
    chr = 1,
    start = 1,
    end = NULL,
    rect_width = 0.3,
    line_width = 0.5,
    arrow_segments = 10,
    arrow_width = 30,
    arrow_length = 0.08,
    arrow_type = "open",
    text_size = 4) {

    .getLevel <- function(txdt) {
        txdt <- txdt[order(start, end), ]

        step <- 1
        while (any(txdt[, set != 1])) {
            x1 <- -1
            for (i in seq_len(nrow(txdt))) {
                if (txdt[i, set == 0 & start > x1]) {
                    txdt[i, level := step]
                    txdt[i, set := 1]
                    x1 <- txdt[i, end]
                }
            }
            step <- step + 1
        }
        return(txdt)
    }


    .getTxdt <- function(dt) {
        transcripts <- dt[, unique(transcript_id)]
        txdt <- data.table::data.table()
        for (i in transcripts) {
            exdt <- dt[transcript_id == i, ]
            txdt <- rbind(txdt,
                data.table::data.table(
                    start = exdt[, min(start)],
                    end = exdt[, max(end)],
                    transcript_id = i,
                    transcript_name = exdt[, unique(transcript_name)],
                    gene_id = exdt[, unique(gene_id)],
                    gene_name = exdt[, unique(gene_name)],
                    strand = exdt[, unique(strand)]))

        }
        txdt[, set := 0]

        txdt <- .getLevel(txdt)
    }


    .transRect <- function(dt, txdt) {
        rdt <- data.table::data.table()
        transcripts <- dt[, unique(transcript_id)]
        for (tx in transcripts) {
            dt[transcript_id == tx,
                level := txdt[transcript_id == tx, level]]
        }

        exons <- dt[type == "exon", ]

        for (i in seq_len(nrow(exons))) {
            x1 <- exons[i, start]
            x2 <- exons[i, end]
            y1 <- exons[i, level] - rect_width
            y2 <- exons[i, level] + rect_width

            rdt <- rbind(rdt,
                data.table::data.table(x1 = x1,
                    x2 = x2,
                    y1 = y1,
                    y2 = y2,
                    exon_number =
                        exons[i, exon_number],
                    transcript_id =
                        exons[i, transcript_id],
                    gene_id = exons[i, gene_id],
                    gene_name =
                        exons[i, gene_name]),
                fill = TRUE)
        }
        return(rdt)
    }


    .transArrow <- function(dt) {
        adt <- data.table::data.table()
        for (i in seq_len(nrow(dt))) {
            mi <- dt[i, start]
            ma <- dt[i, end]

            if (dt[i, strand] == "+") {
                x1 <- mi + (((ma - mi) / arrow_segments) *
                        seq(0, arrow_segments - 1))
                x2 <- ma - (((ma - mi) / arrow_segments) *
                        seq(arrow_segments - 1, 0))
            } else if (dt[i, strand] == "-") {
                x1 <- ma + (((mi - ma) / arrow_segments) *
                        seq(0, arrow_segments - 1))
                x2 <- mi - (((mi - ma) / arrow_segments) *
                        seq(arrow_segments - 1, 0))
            }

            y1 <- rep(dt[i, level], arrow_segments)
            y2 <- y1

            adt <- rbind(adt,
                data.table::data.table(
                    x1 = x1,
                    x2 = x2,
                    y1 = y1,
                    y2 = y2,
                    transcript_id = rep(dt[i, transcript_id],
                        arrow_segments),
                    transcript_name = rep(dt[i, transcript_name],
                        arrow_segments),
                    gene_id = rep(dt[i, gene_id],
                        arrow_segments),
                    gene_name = rep(dt[i, gene_name],
                        arrow_segments)))
        }

        return(adt)
    }


    .transText <- function(dt) {
        tdt <- data.table::data.table()

        for (i in seq_len(nrow(dt))) {
            mi <- dt[i, start]
            ma <- dt[i, end]

            x <- (ma + mi) / 2
            y <- dt[i, level] + 0.4

            tdt <- rbind(tdt,
                data.table::data.table(
                    x = x,
                    y = y,
                    transcript_name = dt[i, transcript_name]))

        }
        return(tdt)
    }


    # convert to data.table
    temp <- data.table::as.data.table(rtracklayer::readGFF(gtfFile))
    if("seqname" %in% colnames(temp)) {
        j <- which(colnames(temp) == "seqname")
        colnames(temp)[i] <- "seqid"
    }
    gtfDt <- temp[type != "gene", c("seqid",
            "type",
            "start",
            "end",
            "strand",
            "gene_biotype",
            "gene_name",
            "exon_number",
            "gene_id",
            "transcript_name",
            "transcript_id")]

    # use new variables to avoid ambiguity
    begin <- start
    ed <- end
    if(is.null(ed)) {
        ed <- max(gtfDt[seqid == chr, end])
    }

    # subset features
    gtfDt <- gtfDt[end >= begin & start <= ed & seqid == chr, ]

    # aggregate transcripts
    txdt <- .getTxdt(gtfDt)

    # get tables for plotting
    rectdt <- .transRect(gtfDt, txdt)
    arrowdt <- .transArrow(txdt)
    textdt <- .transText(txdt)

    # plot
    g <- ggplot2::ggplot() +
        ggplot2::geom_rect(data = rectdt,
            mapping = ggplot2::aes(xmin = x1,
                xmax = x2,
                ymin = y1,
                ymax = y2)) +
        ggplot2::geom_segment(data = arrowdt,
            mapping = ggplot2::aes(x = x1,
                y = y1,
                xend = x2,
                yend = y2),
            size = line_width,
            arrow = ggplot2::arrow(angle = arrow_width,
                length = ggplot2::unit(
                    arrow_length,
                    "inches"),
                type = arrow_type)) +
        ggplot2::geom_text(data = textdt,
            mapping = ggplot2::aes(x = x,
                y = y,
                label = transcript_name),
            size = text_size) +
        .themePublication() +
        ggplot2::theme(axis.title.y = ggplot2::element_blank(),
            axis.text.y = ggplot2::element_blank(),
            axis.ticks.y = ggplot2::element_blank(),
            axis.line.y = ggplot2::element_blank()) +
        ggplot2::xlab(paste0("Chr", chr))

    return(g)
}
